File: solver.c
#include "solver.h"
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

int *read_file(char *path) {
    FILE *fin = fopen(path, "r");
    if (fin == NULL) {
        printf("Error: Can't open file\n");
        exit(2);
    }

    int *sudoku = malloc(sizeof(int) * 81);
    if (sudoku == NULL) {
        printf("Error: Couldn't make board\n");
        exit(3);
    }

    int count = 0;
    char *line = NULL;
    size_t len = 0;
    ssize_t read;
    while ((read = getline(&line, &len, fin)) != -1 && count < 9) {
        if (read != 10) {
            printf("Error: One line contains too many characters\n");
            exit(5);
        }

        for (size_t i = 0; i < 9; i++) {
            if (line[i] - '0' > 9 || line[i] - '0' < 0) {
                printf("Error: Sudoku contains wrong numbers\n");
                exit(4);
            }
            *(sudoku + count * 9 + i) = line[i] - '0';
        }
        count++;
    }

    fclose(fin);
    free(line);
    return sudoku;
}

void print_sudoku(int *sudoku) {
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            printf("%i ", *(sudoku + j + 9 * i));
        }
        printf("\n");
    }
}

bool in_row(int *sudoku, const int row, const int num) {
    for (size_t i = 0; i < 9; i++) {
        if (*(sudoku + row * 9 + i) == num) {
            return true;
        }
    }
    return false;
}

bool in_col(int *sudoku, const int col, const int num) {
    for (size_t i = 0; i < 9; i++) {
        if (*(sudoku + i * 9 + col) == num) {
            return true;
        }
    }
    return false;
}

bool in_box(int *sudoku, const int x, const int y, const int num) {
    int box_x = x - (x % 3);
    int box_y = y - (y % 3);

    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            if (*(sudoku + (box_y + j) * 9 + box_x + i) == num) {
                return true;
            }
        }
    }
    return false;
}

bool find_empty_cell(int *sudoku, int *list) {
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            if (*(sudoku + (i * 9) + j) == 0) {
                list[0] = j;
                list[1] = i;
                return true;
            }
        }
    }
    return false;
}

bool is_safe(int *sudoku, const int x, const int y, const int num) {
    return (
        !in_row(sudoku, y, num) && !in_col(sudoku, x, num) &&
        !in_box(sudoku, x, y, num));
}

bool solve(int *sudoku) {
    int list[2];

    if (!find_empty_cell(sudoku, list)) {
        return true;
    }

    for (int num = 1; num < 10; num++) {
        if (is_safe(sudoku, list[0], list[1], num)) {
            *(sudoku + (list[1] * 9) + list[0]) = num;

            if (solve(sudoku)) {
                return true;
            }

            *(sudoku + (list[1] * 9) + list[0]) = 0;
        }
    }
    return false;
}

File: solver.py
def print_sudoku(sudoku: list[list[int]]) -> None:
    for line in sudoku:
        print(line)

def in_row(sudoku: list[list[int]], row: int, num: int) -> bool:
    for i in range(9):
        if sudoku[row][i] == num:
            return True
    return False


def in_col(sudoku: list[list[int]], col: int, num: int) -> bool:
    for i in range(9):
        if sudoku[i][col] == num:
            return True
    return False


def in_box(sudoku: list[list[int]], x: int, y: int, num: int) -> bool:
    box_x = x - (x % 3) 
    box_y = y - (y % 3)
    for i in range(3):
        for j in range(3):
            if sudoku[box_x + i][box_y + j] == num:
                return True
    return False 

def find_empty_cell(sudoku: list[list[int]], l: list[int]) -> bool:
    for y in range(9):
        for x in range(9):
            if sudoku[x][y] == 0:
                l[0] = x
                l[1] = y
                return True
    return False

def is_safe(sudoku: list[list[int]], x: int, y: int, num: int):
    return (not in_row(sudoku, x, num) and
            not in_col(sudoku, y, num) and
            not in_box(sudoku, x, y, num))

def read_sudoku(file_path: str) -> list[list[int]]:
    with open(file_path, "r") as fin:
        sudoku = []
        lines = fin.readlines()
        for line in lines:
            list = [ord(ch) - ord('0')
                    for ch in line
                    if ord(ch) != ord('\n')]
            sudoku.append(list)
    return sudoku

def solve(sudoku: list[list[int]]) -> bool:
    l = [0, 0]

    if (not find_empty_cell(sudoku, l)):
        return True

    for num in range(1, 10):
        if is_safe(sudoku, l[0], l[1], num):
            sudoku[l[0]][l[1]] = num

            if solve(sudoku):
                return True

            sudoku[l[0]][l[1]] = 0

    return False

def main() -> None:
    sudoku = read_sudoku("./sudoku")
    if solve(sudoku):
        print_sudoku(sudoku)

if __name__ == '__main__':
    main()
